build_test/lib/src/resolve_source.dart

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:build/build.dart';
+import 'package:build_barback/build_barback.dart';
+
+import 'in_memory_reader.dart';
+import 'in_memory_writer.dart';
+import 'multi_asset_reader.dart';
+import 'package:package_resolver/package_resolver.dart';
+import 'package_reader.dart';
+
+/// Returns a future that completes with a source [Resolver] for [inputSource].
+///
+/// Example use:
+/// ```dart
+/// var resolver = await resolveSource(r'''
+///   library example;
+///
+///   import 'dart:collection';
+///
+///   abstract class Foo implements Map {}
+/// ''');
+/// var element = resolver.getLibraryByName('example');
+/// expect(element.getType('Foo'), isNotNull);
+/// ```
+///
+/// **NOTE**: All `package` dependencies are resolved using [PackageAssetReader]
+/// - by default, [PackageAssetReader.currentIsolate]. A custom [resolver] may
+/// be provided to map files not visible to the current package's runtime.
+Future<Resolver> resolveSource(
+  String inputSource, {
+  AssetId inputId,
+  PackageResolver resolver,
+}) async {
+  // If not provided use a fake asset and package.
+  inputId ??= new AssetId('_resolve_source', '_resolve_source.dart');
+  resolver ??= PackageResolver.current;
+  var syncResolver = await resolver.asSync;
+  var reader = new PackageAssetReader(syncResolver, inputId.package);
+  var completer = new Completer<Resolver>();
+  var builder = new _ResolveSourceBuilder(completer);
+  var inputs = [inputId];
+  var inMemory = new InMemoryAssetReader(
+    sourceAssets: {
+      inputId: new DatedString(inputSource),
+    },
+    rootPackage: inputId.package,
+  );
+  // We don't care about the results of this build.
+  // ignore: unawaited_futures
+  runBuilder(
+    builder,
+    inputs,
+    new MultiAssetReader([inMemory, reader]),
+    new InMemoryAssetWriter(),
+    const BarbackResolvers(),
+    rootPackage: inputId.package,
+  );
+  return completer.future;
```

Comment:
Note that `runBuilder` also calls `buildStep.complete` which does some cleanup and nothing (including the Resolver) is guaranteed to work after that point. It seems like since you aren't awaiting the `runBuilder` call there is also a bit of a race condition at play here. The future for the completer might run before `buildStep.complete` actually gets invoked, or it might be later. Any other future you await would probably not finish until after `buildStep.complete` though, which could lead to subtle bugs.

You could potentially fix this by passing a future into the builder, which it awaits on inside its `build` method after completing `completer` with the resolver. You wouldn't complete that future until `tearDown` most likely.